src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/LanguageServices/LanguageServiceHost.cs

```diff
@@ -132,12 +133,10 @@ private Task OnProjectFactoryCompletedAsync()
         {
             var previousProjectContext = _currentAggregateProjectContext;
             var newProjectContext = await UpdateProjectContextAsync().ConfigureAwait(false);
+
             if (previousProjectContext != newProjectContext)
             {
-                // Dispose existing subscriptions.
-                DisposeAndClearSubscriptions();
```

Comment:
Adding the email conversation with @lifengl for more context.

```
From: Manish Vasani 
Sent: Wednesday, February 15, 2017 7:18 PM
To: Lifeng Lu <lifengl@microsoft.com>; Srivatsn Narayanan <srivatsn@microsoft.com>; Jeremy Meng <Jeremy.Meng@microsoft.com>; Jacob Viau <javia@microsoft.com>; Viktor Veis <viveis@microsoft.com>
Cc: David Kean <David.Kean@microsoft.com>
Subject: RE: https://github.com/dotnet/roslyn-project-system/issues/1563

That seems to be indeed the case – I receive one additional update on the disposed subscription, and from then on the updates are only on the newly registered subscription.
Thanks for your help Lifeng.

From: Lifeng Lu 
Sent: Wednesday, February 15, 2017 4:25 PM
To: Manish Vasani <mavasani@microsoft.com>; Srivatsn Narayanan <srivatsn@microsoft.com>; Jeremy Meng <Jeremy.Meng@microsoft.com>; Jacob Viau <javia@microsoft.com>; Viktor Veis <viveis@microsoft.com>
Cc: David Kean <David.Kean@microsoft.com>
Subject: RE: https://github.com/dotnet/roslyn-project-system/issues/1563

Do you only receive a few updates after that, or you will receive new updates forever?

The problem is that dataflow blocks has an internal queue, so even if you remove an subscription, you will still receive data, which has already been added to the queue.
Because your code to handle the dataflow change requires to switch to the UI thread immediately, your handler can be blocked for quite some time, and quite some data will be queued in the action block queue.

From: Manish Vasani 
Sent: Wednesday, February 15, 2017 4:20 PM
To: Lifeng Lu <lifengl@microsoft.com>; Srivatsn Narayanan <srivatsn@microsoft.com>; Jeremy Meng <Jeremy.Meng@microsoft.com>; Jacob Viau <javia@microsoft.com>; Viktor Veis <viveis@microsoft.com>
Cc: David Kean <David.Kean@microsoft.com>
Subject: RE: https://github.com/dotnet/roslyn-project-system/issues/1563

So, my fix is now preventing us from creating duplicate subscriptions for this case.
However, I still have an unanswered CPS question: invoking Dispose on the subscription links doesn’t seem to stop the original subscription handler from receiving CPS updates.
We always dispose all the existing subscriptions here before adding new subscriptions for new set of configured projects, but as I mentioned earlier we still got updates on the original subscription handler.
Am I missing something around how subscription dispose is supposed to work?
```

To summarize, I was assuming that disposing all existing subscriptions and then registering new ones based on new set of active configured projects would avoid duplicate subscriptions. This is not completely true due to the way dataflow queue works, and this caused duplicate updates on the old and new subscription.